File: internal/agent/agent.go
package agent

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/tuist/l10n/internal/checks"
	"github.com/tuist/l10n/internal/config"
	"github.com/tuist/l10n/internal/llm"
	"github.com/tuist/l10n/internal/plan"
	"github.com/tuist/l10n/internal/tools"
)

type Agent struct {
	Client  *llm.Client
	Checker checks.Checker
}

type TranslationRequest struct {
	Source      string
	TargetLang  string
	Format      plan.Format
	Context     string
	Preserve    []string
	Frontmatter string
	CheckCmd    string
	CheckCmds   map[string]string
	Retries     int
	Coordinator config.AgentConfig
	Translator  config.AgentConfig
}

func (a *Agent) Translate(ctx context.Context, req TranslationRequest) (string, error) {
	if a.Client == nil {
		return "", errors.New("llm client is required")
	}

	content := req.Source
	frontmatter := ""
	if req.Format == plan.FormatMarkdown && req.Frontmatter == config.FrontmatterPreserve {
		var ok bool
		frontmatter, content, ok = splitMarkdownFrontmatter(req.Source)
		if !ok {
			content = req.Source
			frontmatter = ""
		}
	}

	brief, err := a.buildBrief(ctx, req)
	if err != nil {
		return "", err
	}

	attempts := req.Retries
	if attempts < 0 {
		attempts = 0
	}
	var lastErr error
	for attempt := 0; attempt <= attempts; attempt++ {
		translation, err := a.translateOnce(ctx, req, brief, content, lastErr)
		if err != nil {
			lastErr = err
			continue
		}

		final := translation
		if isStructuredFormat(req.Format) {
			final = stripCodeFence(final)
		}
		if frontmatter != "" {
			if strings.TrimSpace(final) != "" {
				final = frontmatter + "\n" + final
			} else {
				final = frontmatter + "\n"
			}
		}

		checkErr := a.Checker.Validate(ctx, req.Format, final, req.Source, checks.Options{
			Preserve:  req.Preserve,
			CheckCmd:  req.CheckCmd,
			CheckCmds: req.CheckCmds,
		})
		if checkErr == nil {
			return final, nil
		}
		lastErr = checkErr
	}

	if lastErr == nil {
		lastErr = errors.New("translation failed")
	}
	return "", lastErr
}

func (a *Agent) buildBrief(ctx context.Context, req TranslationRequest) (string, error) {
	model := strings.TrimSpace(req.Coordinator.Model)
	if model == "" {
		return defaultBrief(req), nil
	}

	prompt := fmt.Sprintf(`You are a localization coordinator.
Create a short translation brief for the translator.
The brief must be plain text and under 12 lines.

Target language: %s
Format: %s
Preserve: %s
Frontmatter mode: %s
Tools: %s

Context:
%s
`, req.TargetLang, req.Format, strings.Join(req.Preserve, ", "), req.Frontmatter, tools.Summary(), req.Context)

	resp, err := a.Client.Chat(ctx, req.Coordinator, model, []llm.ChatMessage{
		{Role: "system", Content: "You coordinate translations and produce concise briefs."},
		{Role: "user", Content: prompt},
	})
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(resp), nil
}

func defaultBrief(req TranslationRequest) string {
	lines := []string{
		"Translate the content faithfully and naturally.",
		"Preserve code blocks, inline code, URLs, and placeholders.",
		"Keep formatting, lists, and headings intact.",
		"Return only the translated content.",
	}
	if isStructuredFormat(req.Format) {
		lines = append(lines, "Return valid "+string(req.Format)+" only. Do not wrap in markdown fences.")
	}
	if req.Frontmatter == config.FrontmatterPreserve {
		lines = append(lines, "Frontmatter is preserved separately; do not add new frontmatter.")
	}
	lines = append(lines, "Tools run after translation: "+tools.Summary()+".")
	return strings.Join(lines, "\n")
}

func (a *Agent) translateOnce(ctx context.Context, req TranslationRequest, brief string, content string, lastErr error) (string, error) {
	model := strings.TrimSpace(req.Translator.Model)
	if model == "" {
		return "", errors.New("translator model is required")
	}

	user := fmt.Sprintf("Translate to %s.\n\nContext:\n%s\n\nSource:\n%s", req.TargetLang, req.Context, content)
	if lastErr != nil {
		user += fmt.Sprintf("\n\nPrevious output failed validation: %s\nReturn a corrected full translation.", lastErr.Error())
	}

	resp, err := a.Client.Chat(ctx, req.Translator, model, []llm.ChatMessage{
		{Role: "system", Content: fmt.Sprintf("You are a translation engine. Follow this brief:\n%s", brief)},
		{Role: "user", Content: user},
	})
	if err != nil {
		return "", err
	}
	return strings.TrimRight(resp, "\n"), nil
}

func splitMarkdownFrontmatter(contents string) (frontmatter string, body string, ok bool) {
	lines := strings.Split(contents, "\n")
	if len(lines) == 0 {
		return "", contents, false
	}
	marker := strings.TrimSpace(lines[0])
	if marker != "---" && marker != "+++" {
		return "", contents, false
	}

	end := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == marker {
			end = i
			break
		}
	}
	if end == -1 {
		return "", contents, false
	}

	frontmatter = strings.Join(lines[:end+1], "\n")
	body = strings.Join(lines[end+1:], "\n")
	return frontmatter, body, true
}

func isStructuredFormat(format plan.Format) bool {
	switch format {
	case plan.FormatJSON, plan.FormatYAML, plan.FormatPO:
		return true
	default:
		return false
	}
}

func stripCodeFence(content string) string {
	trimmed := strings.TrimSpace(content)
	if !strings.HasPrefix(trimmed, "```") {
		return content
	}
	lines := strings.Split(trimmed, "\n")
	if len(lines) < 2 {
		return content
	}
	if !strings.HasPrefix(strings.TrimSpace(lines[0]), "```") {
		return content
	}
	if strings.TrimSpace(lines[len(lines)-1]) != "```" {
		return content
	}
	return strings.Join(lines[1:len(lines)-1], "\n")
}

File: src/agent.ts
import type { AgentConfig } from "./config.js";
import { FRONTMATTER_PRESERVE } from "./config.js";
import { chat, type ChatMessage, type TranslateGemmaContentPart } from "./llm.js";
import { validate, type CheckOptions } from "./checks.js";
import type { Format } from "./format.js";
import type { Reporter } from "./reporter.js";
import { toolsSummary } from "./tools.js";

export interface TranslationRequest {
  source: string;
  targetLang: string;
  format: Format;
  context: string;
  preserve: string[];
  frontmatter: string;
  checkCmd: string;
  checkCmds: Record<string, string>;
  toolReporter?: Reporter;
  progressLabel: string;
  progressCurrent: number;
  progressTotal: number;
  retries: number;
  coordinator: AgentConfig;
  translator: AgentConfig;
  root: string;
}

export async function translate(req: TranslationRequest): Promise<string> {
  let content = req.source;
  let frontmatter = "";

  if (req.format === "markdown" && req.frontmatter === FRONTMATTER_PRESERVE) {
    const split = splitMarkdownFrontmatter(req.source);
    if (split.ok) {
      frontmatter = split.frontmatter;
      content = split.body;
    }
  }

  const brief = await buildBrief(req);

  let attempts = req.retries;
  if (attempts < 0) attempts = 0;
  let lastErr: Error | null = null;

  for (let attempt = 0; attempt <= attempts; attempt++) {
    let translation: string;
    try {
      translation = await translateOnce(req, brief, content, lastErr);
    } catch (err: any) {
      lastErr = err;
      continue;
    }

    let final = translation;
    if (isStructuredFormat(req.format)) {
      final = stripCodeFence(final);
    }
    if (frontmatter) {
      if (final.trim()) {
        final = frontmatter + "\n" + final;
      } else {
        final = frontmatter + "\n";
      }
    }

    try {
      await validate(req.root, req.format, final, req.source, {
        preserve: req.preserve,
        checkCmd: req.checkCmd,
        checkCmds: req.checkCmds,
        reporter: req.toolReporter,
        label: req.progressLabel,
        current: req.progressCurrent,
        total: req.progressTotal,
      });
      return final;
    } catch (checkErr: any) {
      lastErr = checkErr;
    }
  }

  throw lastErr ?? new Error("translation failed");
}

async function buildBrief(req: TranslationRequest): Promise<string> {
  const model = (req.coordinator.model ?? "").trim();
  if (!model) return defaultBrief(req);

  const prompt = `You are a localization coordinator.
Create a short translation brief for the translator.
The brief must be plain text and under 12 lines.

Target language: ${req.targetLang}
Format: ${req.format}
Preserve: ${req.preserve.join(", ")}
Frontmatter mode: ${req.frontmatter}
Tools: ${toolsSummary()}

Context:
${req.context}
`;

  const resp = await chat(req.coordinator, model, [
    {
      role: "system",
      content: "You coordinate translations and produce concise briefs.",
    },
    { role: "user", content: prompt },
  ]);
  return resp.trim();
}

function defaultBrief(req: TranslationRequest): string {
  const lines = [
    "Translate the content faithfully and naturally.",
    "Preserve code blocks, inline code, URLs, and placeholders.",
    "Keep formatting, lists, and headings intact.",
    "Return only the translated content.",
  ];
  if (isStructuredFormat(req.format)) {
    lines.push(`Return valid ${req.format} only. Do not wrap in markdown fences.`);
  }
  if (req.frontmatter === FRONTMATTER_PRESERVE) {
    lines.push("Frontmatter is preserved separately; do not add new frontmatter.");
  }
  lines.push(`Tools run after translation: ${toolsSummary()}.`);
  return lines.join("\n");
}

// TranslateGemma uses ISO 639-1 codes; map l10n's BCP 47 variants to what the model expects.
const LANG_CODE_MAP: Record<string, string> = {
  "zh-Hans": "zh",
  "zh-Hant": "zh",
};

function mapLangCode(lang: string): string {
  return LANG_CODE_MAP[lang] ?? lang;
}

function isTranslateGemma(model: string): boolean {
  return model.toLowerCase().includes("translategemma");
}

async function translateOnce(
  req: TranslationRequest,
  brief: string,
  content: string,
  lastErr: Error | null,
): Promise<string> {
  const model = (req.translator.model ?? "").trim();
  if (!model) throw new Error("translator model is required");

  if (isTranslateGemma(model)) {
    return translateOnceGemma(req, model, content);
  }

  let user = `Translate to ${req.targetLang}.\n\nContext:\n${req.context}\n\nSource:\n${content}`;
  if (lastErr) {
    user += `\n\nPrevious output failed validation: ${lastErr.message}\nReturn a corrected full translation.`;
  }

  const resp = await chat(req.translator, model, [
    {
      role: "system",
      content: `You are a translation engine. Follow this brief:\n${brief}`,
    },
    { role: "user", content: user },
  ]);
  return resp.replace(/\n+$/, "");
}

async function translateOnceGemma(
  req: TranslationRequest,
  model: string,
  content: string,
): Promise<string> {
  const parts: TranslateGemmaContentPart[] = [
    {
      type: "text",
      source_lang_code: "en",
      target_lang_code: mapLangCode(req.targetLang),
      text: content,
    },
  ];

  const resp = await chat(req.translator, model, [{ role: "user", content: parts }]);
  return resp.replace(/\n+$/, "");
}

function splitMarkdownFrontmatter(contents: string): {
  frontmatter: string;
  body: string;
  ok: boolean;
} {
  const lines = contents.split("\n");
  if (lines.length === 0) return { frontmatter: "", body: contents, ok: false };
  const marker = lines[0].trim();
  if (marker !== "---" && marker !== "+++") {
    return { frontmatter: "", body: contents, ok: false };
  }

  let end = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === marker) {
      end = i;
      break;
    }
  }
  if (end === -1) return { frontmatter: "", body: contents, ok: false };

  const frontmatter = lines.slice(0, end + 1).join("\n");
  const body = lines.slice(end + 1).join("\n");
  return { frontmatter, body, ok: true };
}

function isStructuredFormat(format: Format): boolean {
  return format === "json" || format === "yaml" || format === "po";
}

function stripCodeFence(content: string): string {
  const trimmed = content.trim();
  if (!trimmed.startsWith("```")) return content;
  const lines = trimmed.split("\n");
  if (lines.length < 2) return content;
  if (!lines[0].trim().startsWith("```")) return content;
  if (lines[lines.length - 1].trim() !== "```") return content;
  return lines.slice(1, lines.length - 1).join("\n");
}
